File: AscentSchoolsMobile/app/src/main/java/com/ascentschools/mobile/util/ReceiptPrinter.kt
package com.ascentschools.mobile.util

import android.content.Context
import android.print.PrintAttributes
import android.print.PrintManager
import android.view.View
import android.webkit.WebView
import android.webkit.WebViewClient
import com.ascentschools.mobile.data.api.ReceiptDetailDto

/**
 * Renders a fee receipt to an offscreen WebView and hands it to Android's
 * PrintManager — the user gets the system print sheet (real printer OR
 * "Save as PDF"). No third-party PDF library required.
 */
object ReceiptPrinter {

    // Hold the WebView so it isn't garbage-collected before the print adapter runs.
    private var webViewRef: WebView? = null

    fun print(context: Context, receipt: ReceiptDetailDto, schoolName: String?) {
        val wv = WebView(context)
        // Offscreen print-only WebView: force software rendering so it never
        // touches the GPU/Vulkan path. Avoids native crashes inside the device
        // GPU driver (Adreno vkCreateInstance) during WebView GPU init, and is
        // the documented recommendation for WebViews used purely for printing.
        wv.setLayerType(View.LAYER_TYPE_SOFTWARE, null)
        wv.webViewClient = object : WebViewClient() {
            override fun onPageFinished(view: WebView, url: String?) {
                val jobName = "Receipt_" + (receipt.receiptNo ?: receipt.receiptId.toString())
                val pm = context.getSystemService(Context.PRINT_SERVICE) as PrintManager
                val adapter = view.createPrintDocumentAdapter(jobName)
                pm.print(jobName, adapter, PrintAttributes.Builder().build())
                webViewRef = null
            }
        }
        wv.loadDataWithBaseURL(null, buildHtml(receipt, schoolName), "text/html", "UTF-8", null)
        webViewRef = wv
    }

    private fun money(v: Double) = "₹" + String.format("%.2f", v)

    private fun esc(s: String?): String = (s ?: "")
        .replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")

    private fun buildHtml(r: ReceiptDetailDto, schoolName: String?): String {
        val rows = StringBuilder()
        for (it in r.items) {
            val name = it.feeTypeName ?: it.routeName ?: it.hostelName ?: "-"
            rows.append(
                "<tr>" +
                    "<td>${esc(name)}</td>" +
                    "<td>${esc(it.termName ?: "-")}</td>" +
                    "<td class='r'>${money(it.amount)}</td>" +
                    "<td class='r'>${if (it.concessionAmount > 0) money(it.concessionAmount) else "-"}</td>" +
                    "<td class='r'>${money(it.netAmount)}</td>" +
                "</tr>"
            )
        }

        val cancelled = (r.status ?: "").equals("Cancelled", ignoreCase = true)
        val cancelBanner = if (cancelled)
            "<div class='cancel'>CANCELLED</div>" else ""

        return """
            <html><head><meta name="viewport" content="width=device-width, initial-scale=1"/>
            <style>
              body { font-family: -apple-system, Roboto, Arial, sans-serif; color:#222; padding:16px; }
              .hdr { text-align:center; border-bottom:2px solid #1677ff; padding-bottom:8px; margin-bottom:12px; }
              .hdr h2 { margin:0; color:#1677ff; }
              .hdr .sub { color:#555; font-size:13px; }
              .meta { display:flex; justify-content:space-between; font-size:13px; margin-bottom:10px; }
              table { width:100%; border-collapse:collapse; font-size:13px; }
              th,td { border:1px solid #ddd; padding:6px 8px; text-align:left; }
              th { background:#f5f7fa; }
              .r { text-align:right; }
              .total { text-align:right; font-weight:bold; font-size:15px; margin-top:10px; }
              .foot { margin-top:24px; font-size:12px; color:#666; display:flex; justify-content:space-between; }
              .cancel { color:#cf1322; border:2px solid #cf1322; display:inline-block; padding:2px 10px;
                        font-weight:bold; transform:rotate(-6deg); margin-top:6px; }
            </style></head><body>
              <div class="hdr">
                <h2>${esc(schoolName ?: "Fee Receipt")}</h2>
                <div class="sub">Fee Receipt</div>
                $cancelBanner
              </div>
              <div class="meta">
                <div>
                  <div><b>Receipt No:</b> ${esc(r.receiptNo ?: r.receiptId.toString())}</div>
                  <div><b>Date:</b> ${esc(r.paymentDate ?: "")}</div>
                  <div><b>Mode:</b> ${esc(r.paymentModeName ?: "")}</div>
                </div>
                <div>
                  <div><b>Student:</b> ${esc(r.studentName ?: "")}</div>
                  <div><b>Adm No:</b> ${esc(r.admissionNo ?: "")}</div>
                  <div><b>Class:</b> ${esc(r.className ?: "")} &nbsp; ${esc(r.academicYear ?: "")}</div>
                </div>
              </div>
              <table>
                <thead><tr><th>Fee</th><th>Term</th><th class="r">Amount</th>
                  <th class="r">Concession</th><th class="r">Net</th></tr></thead>
                <tbody>$rows</tbody>
              </table>
              <div class="total">Total Paid: ${money(r.totalAmount)}</div>
              <div class="foot">
                <span>Collected by: ${esc(r.createdBy ?: "")}</span>
                <span>This is a computer-generated receipt.</span>
              </div>
            </body></html>
        """.trimIndent()
    }
}

File: ascentschooliosapp/ascentschooliosapp/Features/Parent/Fees/ReceiptPrinter.swift
import UIKit

// Phase 58 (Android parity): in-app receipt print. Android uses an offscreen
// WebView + PrintManager; iOS collapses those to UIPrintInteractionController
// with a UIMarkupTextPrintFormatter, which shows the system print sheet (also
// doubles as a preview + "Save to PDF" affordance) with zero WKWebView setup.
//
// Called from FeesView after the view model finishes fetching the receipt
// detail. Presentation must happen on the main actor from a live UIWindow —
// SwiftUI doesn't own the print controller, so we grab the topmost window ourselves.
@MainActor
enum ReceiptPrinter {

    static func present(receipt: MobileReceiptDto) {
        let html = renderHTML(receipt: receipt)

        let info = UIPrintInfo.printInfo()
        info.outputType = .general
        info.jobName = "Receipt \(receipt.receiptNo ?? String(receipt.receiptId))"
        info.orientation = .portrait

        let controller = UIPrintInteractionController.shared
        controller.printInfo = info
        controller.printFormatter = UIMarkupTextPrintFormatter(markupText: html)
        controller.showsNumberOfCopies = true

        guard let window = topWindow() else {
            controller.present(animated: true, completionHandler: nil)
            return
        }

        // iPad requires a source rect/bar button for the popover presentation;
        // iPhone doesn't care. Pass a centred zero-rect on iPad so the sheet
        // anchors reliably even without a tapped-button reference.
        if UIDevice.current.userInterfaceIdiom == .pad, let root = window.rootViewController {
            let sourceRect = CGRect(
                x: root.view.bounds.midX,
                y: root.view.bounds.midY,
                width: 0,
                height: 0
            )
            controller.present(from: sourceRect, in: root.view, animated: true, completionHandler: nil)
        } else {
            controller.present(animated: true, completionHandler: nil)
        }
    }

    // MARK: - HTML

    private static func renderHTML(receipt: MobileReceiptDto) -> String {
        let brand = AppInfo.displayName.htmlEscaped
        let receiptNo = (receipt.receiptNo ?? "—").htmlEscaped
        let receiptDate = formatDate(receipt.receiptDate)
        let studentName = (receipt.studentName ?? "—").htmlEscaped
        let admission = (receipt.admissionNo ?? "—").htmlEscaped
        let classSec = classSection(className: receipt.className, sectionName: receipt.sectionName)
        let year = (receipt.academicYear ?? "").htmlEscaped
        let mode = (receipt.paymentMode ?? "—").htmlEscaped
        let collectedBy = (receipt.createdBy ?? "—").htmlEscaped
        let isCancelled = (receipt.status ?? "").caseInsensitiveCompare("Cancelled") == .orderedSame

        let itemRows = receipt.items.map { item in
            let feeType = firstNonEmpty(item.feeTypeName, item.routeName, item.hostelName) ?? "—"
            let term = firstNonEmpty(item.termName, item.feePeriodLabel) ?? "—"
            return """
            <tr>
              <td>\(feeType.htmlEscaped)</td>
              <td>\(term.htmlEscaped)</td>
              <td class="num">\(rupees(item.amount))</td>
              <td class="num">\(rupees(item.concessionAmount))</td>
              <td class="num">\(rupees(item.netAmount))</td>
            </tr>
            """
        }.joined(separator: "\n")

        let cancelledBanner = isCancelled
            ? """
              <div class="cancelled">CANCELLED\(receipt.remarks.map { " · \($0.htmlEscaped)" } ?? "")</div>
              """
            : ""

        return """
        <!DOCTYPE html>
        <html>
        <head>
          <meta charset="utf-8" />
          <title>Fee Receipt</title>
          <style>
            body { font-family: -apple-system, "Helvetica Neue", Helvetica, Arial, sans-serif;
                   color: #0F172A; margin: 24px; }
            .header { text-align: center; margin-bottom: 16px; }
            .school { font-size: 20px; font-weight: 700; margin: 0; }
            .title { font-size: 14px; letter-spacing: 2px; color: #64748B;
                     margin-top: 6px; text-transform: uppercase; }
            .meta { width: 100%; margin-top: 16px; font-size: 12px; }
            .meta td { padding: 4px 0; vertical-align: top; }
            .meta .label { color: #64748B; padding-right: 12px; width: 130px; }
            table.items { width: 100%; border-collapse: collapse; margin-top: 18px; font-size: 12px; }
            table.items th, table.items td { border-bottom: 1px solid #E2E8F0; padding: 8px 6px;
                                             text-align: left; }
            table.items th { background: #F1F5F9; color: #1E3A8A; font-weight: 600; }
            table.items td.num, table.items th.num { text-align: right; }
            .total-row td { border-top: 2px solid #1E3A8A; border-bottom: none;
                            font-weight: 700; font-size: 13px; padding-top: 10px; }
            .cancelled { color: #B91C1C; border: 2px solid #B91C1C; padding: 8px 12px;
                         font-weight: 700; margin-top: 14px; text-align: center; letter-spacing: 1px; }
            .footer { margin-top: 24px; font-size: 11px; color: #64748B; text-align: center; }
          </style>
        </head>
        <body>
          <div class="header">
            <p class="school">\(brand)</p>
            <p class="title">Fee Receipt</p>
          </div>

          \(cancelledBanner)

          <table class="meta">
            <tr><td class="label">Receipt No.</td><td>\(receiptNo)</td>
                <td class="label">Date</td><td>\(receiptDate)</td></tr>
            <tr><td class="label">Student</td><td>\(studentName)</td>
                <td class="label">Admission No.</td><td>\(admission)</td></tr>
            <tr><td class="label">Class</td><td>\(classSec)</td>
                <td class="label">Academic Year</td><td>\(year)</td></tr>
            <tr><td class="label">Payment Mode</td><td>\(mode)</td>
                <td class="label">Collected By</td><td>\(collectedBy)</td></tr>
          </table>

          <table class="items">
            <thead>
              <tr>
                <th>Fee Type</th>
                <th>Term / Period</th>
                <th class="num">Amount</th>
                <th class="num">Concession</th>
                <th class="num">Net</th>
              </tr>
            </thead>
            <tbody>
              \(itemRows)
              <tr class="total-row">
                <td colspan="4">Total</td>
                <td class="num">\(rupees(receipt.totalAmount))</td>
              </tr>
            </tbody>
          </table>

          <div class="footer">Generated in-app on \(currentTimestamp()) · \(brand)</div>
        </body>
        </html>
        """
    }

    // MARK: - Helpers

    private static func rupees(_ value: Double) -> String {
        let f = NumberFormatter()
        f.numberStyle = .decimal
        f.minimumFractionDigits = 2
        f.maximumFractionDigits = 2
        let text = f.string(from: NSNumber(value: value)) ?? String(format: "%.2f", value)
        return "₹" + text
    }

    private static func formatDate(_ iso: String?) -> String {
        guard let iso, !iso.isEmpty else { return "—" }
        let display = DateFormatter()
        display.dateFormat = "d MMM yyyy"

        let isoFull = ISO8601DateFormatter()
        isoFull.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let date = isoFull.date(from: iso) { return display.string(from: date) }

        let isoNoFraction = ISO8601DateFormatter()
        isoNoFraction.formatOptions = [.withInternetDateTime]
        if let date = isoNoFraction.date(from: iso) { return display.string(from: date) }

        let dateOnly = DateFormatter()
        dateOnly.dateFormat = "yyyy-MM-dd"
        if let date = dateOnly.date(from: iso) { return display.string(from: date) }

        return iso
    }

    private static func currentTimestamp() -> String {
        let f = DateFormatter()
        f.dateFormat = "d MMM yyyy, h:mm a"
        return f.string(from: Date())
    }

    private static func classSection(className: String?, sectionName: String?) -> String {
        let base = (className ?? "").trimmingCharacters(in: .whitespaces)
        let sec = (sectionName ?? "").trimmingCharacters(in: .whitespaces)
        if base.isEmpty && sec.isEmpty { return "—" }
        if sec.isEmpty { return base.htmlEscaped }
        return "\(base) · \(sec)".htmlEscaped
    }

    private static func firstNonEmpty(_ values: String?...) -> String? {
        for v in values {
            if let v, !v.trimmingCharacters(in: .whitespaces).isEmpty { return v }
        }
        return nil
    }

    private static func topWindow() -> UIWindow? {
        let scenes = UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
        return scenes
            .flatMap { $0.windows }
            .first(where: { $0.isKeyWindow }) ?? scenes.flatMap { $0.windows }.first
    }
}

// Minimal HTML escaper — avoids pulling in a third-party dependency for the
// four characters that actually matter in receipt fields.
private extension String {
    var htmlEscaped: String {
        var s = self
        s = s.replacingOccurrences(of: "&", with: "&amp;")
        s = s.replacingOccurrences(of: "<", with: "&lt;")
        s = s.replacingOccurrences(of: ">", with: "&gt;")
        s = s.replacingOccurrences(of: "\"", with: "&quot;")
        s = s.replacingOccurrences(of: "'", with: "&#39;")
        return s
    }
}
